### PR TITLE
Digital: Adds `Logic` and defines 9 logic levels, `StdULogicVector`, and `StdLogicVector`

### PR DESCRIPTION
**Pull request by ven-k**
First in series of PRs to implement the entire IEEE.Std_logic_1164 in Julia

- [x] Adds Logic and its 9 levels
- [x] Accordingly refactors AndTable, OrTable, NotTable, XorTable
- [x] Accordingly refactors _and, _or, _not, _xor
- [x] Adds StdULogicVector and few predefined vectors and StdLogicVector
- [x] Adds relevant tests

----- 
This PR will be followed up to add things like:
- Resolution table
- Inter-conversions from one range of logic-levels to other range
- Digital sources, clocks
- High level DiscreteSystems of gates
- And then revival of Adders, En(/De)coders, (De)Muxes

### CODE REVIEW

File: src/Electrical/Digital/logic.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# Aliases for 9 Digital Logic levels
+const Uninitialized = Logic(1)
+const ForcingUnknown = Logic(2)
+const ForcingZero = Logic(3)
+const ForcingOne = Logic(4)
+const HighImpedence = Logic(5)
+const WeakUnknown = Logic(6)
+const WeakZero = Logic(7)
+const WeakOne = Logic(8)
+const DontCare = Logic(9)
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic.jl:38]: This feels an awful lot like an `@enum`. Have you considered using an enumeration for this type instead?

Comment (ven-k) [reply src/Electrical/Digital/logic.jl:38]: Initially I was considering to include 1 and 0 as is in the set of logic levels; as @enum wouldn't work with that, I dropped the idea then.

Now because of F0, F1 it is doable - Logic as an Enum of these 9 vals.
ATM I can't think of any other field Logic could have. So I will swap this part with an Enum

Comment (ven-k) [reply src/Electrical/Digital/logic.jl:38]: `getfield.(<big StructLogic matrix>, :levels)` vs `Int.(<big EnumLogic matrix>)`
Speeds were very close, and the former was slightly faster. However, the latter always allocated less.

File: src/Electrical/Electrical.jl [hunk 2]
```diff
@@ -26,14 +25,22 @@ include("Analog/sensors.jl")
...
 # TODO:
 # - digital
 # - machines
 # - multi-phase
 
+export Logic
+include("Digital/logic.jl")
+
+export StdLogicVector, StdULogicVector,
+       std_ulogic, UX01, UX01Z, X01, X01Z
```
Comment (baggepinnen) [inline src/Electrical/Electrical.jl:41]: any particular reason for exporting `UX01, UX01Z, X01, X01Z` while not exporting many other vars?

File: src/Electrical/Digital/logic.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    elseif Int(l) == 6
+        print(io, "W")
+    elseif Int(l) == 7
+        print(io, "L")
+    elseif Int(l) == 8
+        print(io, "H")
+    elseif Int(l) == 9
+        print(io, "DC")
+    else
+        print(io, "Invalid logic level")
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic.jl:33]:
```suggestion
        print(io, "Invalid logic level: $l")
```

File: src/Electrical/Digital/logic.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    end
+end
+
+Base.zero(l::Logic) = F0
+Base.zero(l::Type{Logic}) = F0
+Base.one(l::Logic) = F1
+Base.one(l::Type{Logic}) = F1
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic.jl:40]:
```suggestion
Base.zero(::Logic) = F0
Base.zero(::Type{Logic}) = F0
Base.one(::Logic) = F1
Base.one(::Type{Logic}) = F1
```
to avoid potential linter warnings against unused arguments

File: src/Electrical/Digital/logic.jl [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    elseif i == one(i)
+        one(l)
+    else
+        throw("$i isn't a valid `Logic` value")
+    end
+end
+Base.convert(l::Type{Logic}, i::Logic) = i
+
+convert_to_logic(i::Number) = convert(Logic, i)
+convert_to_logic(i::Logic) = i
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic.jl:55]: why are these required when the `Base.convert(l::Type{Logic}, i::Number)` method exists?

File: src/Electrical/Digital/logic.jl [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+get_logic_level(l::Logic) = Int(l)
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic.jl:57]: Is this function really required when `Int` is equally simple?

File: src/Electrical/Digital/logic_vectors.jl [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import Base: size, axes, getindex
+
+const LogicOrNumber = Union{Logic, Number}
+
+struct StdULogicVector{N} <: AbstractArray{Logic, N}
+    logic::AbstractArray{Logic}
+    level::AbstractArray{Int}
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic_vectors.jl:7]: I notice that there's an abstract type here. I realize that MTK will butcher most type stability anyways and that this may not be a problem in this case, but have the implications of this been thought through?

File: src/Electrical/Digital/logic_vectors.jl [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    function StdLogicVector(l::AbstractArray)
+        N = ndims(l)
+        l = AbstractArray{Logic}(convert.(Logic, l))
+        new{N}(l, get_logic_level.(l))
+    end
+end
+
+const LogicVector = Union{StdULogicVector, StdLogicVector}
+
+Base.size(l::L) where {L <: LogicVector} = size(l.logic)
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic_vectors.jl:27]:
```suggestion
Base.size(l::LogicVector) = size(l.logic)
```
this is against the Julia style guide ;)

File: src/Electrical/Digital/logic_vectors.jl [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+Base.axes(l::L) where {L <: LogicVector} = axes(l.logic)
```
Comment (baggepinnen) [inline src/Electrical/Digital/logic_vectors.jl:29]:
```suggestion
Base.axes(l::LogicVector) = axes(l.logic)
```
Same here, the type parameter should really only be introduced if you need to use it (or if you need to force specialization of a function type in some special cases).

File: src/Electrical/Digital/logic_vectors.jl (hunk 8 above)
Comment (baggepinnen) [reply src/Electrical/Digital/logic_vectors.jl:27]: There are several similar instances below. These give precompilation warnings and are best avoided

File: src/Electrical/Digital/tables.jl [hunk 10]
```diff
@@ -1,131 +1,97 @@
-# 9-level logic
-@parameters X
+# AND gate
+AndTable = [
```
Comment (baggepinnen) [inline src/Electrical/Digital/tables.jl:2]: Should this be `const`?

File: src/Electrical/Digital/tables.jl [hunk 11]
```diff
@@ -1,131 +1,97 @@
...
-    i, j = logic(a), logic(b)
-    (typeof(i) != Int || typeof(j) != Int) && return X
-    OrTable[i, j]
+_not(x::Logic) = NotTable[get_logic_level(x)]
+_not(x::Number) = _not(convert_to_logic(x))
+
+@register_symbolic _not(x)
+
+# OR gate
+OrTable = [
```
Comment (baggepinnen) [inline src/Electrical/Digital/tables.jl:40]: const?

File: src/Electrical/Digital/tables.jl [hunk 12]
```diff
@@ -1,131 +1,97 @@
...
+_xor2(a::Number, b::Number) = _xor2(convert_to_logic(a), convert_to_logic(b))
 
 function _xor(x...)
-    y = [_xor2(x[1], 0)]
-    for i in 2:length(x)
+    y = [convert_to_logic(x[1])]
+    for i in 2:lastindex(x)
         push!(y, _xor2(x[i], y[i - 1]))
     end
     return y[end]
```
Comment (baggepinnen) [inline src/Electrical/Digital/tables.jl:94]: You're only ever using the last value of `y`, so do you really need to allocate an array?

File: src/Electrical/Electrical.jl (hunk 2 above)
Comment (ven-k) [reply src/Electrical/Electrical.jl:41]: The IEEE Standard defines these logic vectors.

About the others:
- I expect the logic tables (ex: AndTable) to be used only internally and the logic gates will be exposed to users via (yet-to-be-defined) higher level gate components.
- In the "Standard", Logic values (ex: Uninitialized) are not defined in their full form; but via short form (ex: U, X) and so on. We could export the full form; but usage of short form would be more main stream (; however one could argue that we are deviating with F1, F0 and DC which are simply `1`, `0`, `-`)

File: src/Electrical/Digital/tables.jl (hunk 12 above)
Comment (ven-k) [reply src/Electrical/Digital/tables.jl:94]: Good point

File: src/Electrical/Digital/logic.jl (hunk 6 above)
Comment (ven-k) [reply src/Electrical/Digital/logic.jl:57]: Int(<a logic value>) doesn't really clarify what it is returning or what the intent is.

File: src/Electrical/Digital/logic.jl (hunk 5 above)
Comment (ven-k) [reply src/Electrical/Digital/logic.jl:55]: IMO it would be a convenience function; Could be dropped.

File: src/Electrical/Digital/logic_vectors.jl (hunk 7 above)
Comment (ven-k) [reply src/Electrical/Digital/logic_vectors.jl]: Now this is changed to `Array`

**Review by baggepinnen — APPROVED** [review]

File: src/Electrical/Digital/logic_vectors.jl [hunk 13]
```diff
@@ -0,0 +1,57 @@
+import Base: size, axes, getindex, setindex!
+
+const LogicOrNumber = Union{Logic, Number}
+
+struct StdULogicVector{N} <: AbstractArray{Logic, N}
+    logic::Array{Logic}
```
Comment (YingboMa) [inline src/Electrical/Digital/logic_vectors.jl:6]: Shouldn't it be `Array{Logic, N}`?

File: src/Electrical/Digital/logic_vectors.jl [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+Base.size(l::LogicVector) = size(l.logic)
+
+Base.axes(l::LogicVector) = axes(l.logic)
+
+function Base.getindex(s::LogicVector,
+                       i::Int)
+    getindex(s.logic, i)
+end
+function Base.getindex(s::LogicVector, i1::Int, i2::Int,
```
Comment (YingboMa) [inline src/Electrical/Digital/logic_vectors.jl:35]: Why do we call it a vector if it can be multidimensional?

File: src/Electrical/Digital/logic_vectors.jl [hunk 15]
```diff
@@ -0,0 +1,57 @@
+import Base: size, axes, getindex, setindex!
+
+const LogicOrNumber = Union{Logic, Number}
+
+struct StdULogicVector{N} <: AbstractArray{Logic, N}
+    logic::Array{Logic}
+    level::Array{Int}
```
Comment (YingboMa) [inline src/Electrical/Digital/logic_vectors.jl:7]: Why do we need to store `level` at all if it can be trivially computed from `logic`?

File: src/Electrical/Digital/logic_vectors.jl [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    function StdLogicVector(l::Array)
+        N = ndims(l)
+        l = Array{Logic}(convert.(Logic, l))
+        new{N}(l, get_logic_level.(l))
+    end
+end
+
+const LogicVector = Union{StdULogicVector, StdLogicVector}
```
Comment (YingboMa) [inline src/Electrical/Digital/logic_vectors.jl:25]: What are they used for?

File: src/Electrical/Digital/logic_vectors.jl (hunk 14 above)
Comment (ven-k) [reply src/Electrical/Digital/logic_vectors.jl:35]: To remain consistent with `std_logic_vector` and `std_ulogic_vector` are the names used in IEEE Std_logic_1164.
Although its mostly used in vector form, it can be a matrix as well.

File: src/Electrical/Digital/logic_vectors.jl (hunk 15 above)
Comment (YingboMa) [reply src/Electrical/Digital/logic_vectors.jl:7]: Should we have a logic table so that it takes `Logic` indices, then?

File: src/Electrical/Digital/logic_vectors.jl (hunk 16 above)
Comment (ven-k) [reply src/Electrical/Digital/logic_vectors.jl:25]: StdULogicVector, StdLogicVector are part the Std_logic_1164.
LogicVector is only for convenience while writing dispatches.

File: src/Electrical/Digital/logic_vectors.jl (hunk 13 above)
Comment (ven-k) [reply src/Electrical/Digital/logic_vectors.jl]: Done

File: src/Electrical/Digital/logic_vectors.jl (hunk 15 above)
Comment (YingboMa) [reply src/Electrical/Digital/logic_vectors.jl]: This point is still not addressed. We need `LogicTable{N} <: AbstractArray{Logic, N}` such that we can index into it using `Logic` indices, so we don't even need to store `level`.

Comment (ven-k) [reply src/Electrical/Digital/logic_vectors.jl]: This seemed like the simplest approach to index the values.
I see the point in a `LogicTable` type. Will implement that.

**Review by YingboMa — APPROVED** [review]
